log/scope.go

```diff
@@ -39,10 +38,37 @@ type Scope struct {
 	outputLevel     atomic.Value
 	stackTraceLevel atomic.Value
 	logCallers      atomic.Value
+
+	// context data - key slice to preserve ordering
+	contextKey []string
+	context    map[string]interface{}
+	// labels data - key slice to preserve ordering
+	labelKeys []string
+	labels    map[string]interface{}
 }
 
-var scopes = make(map[string]*Scope)
-var lock = sync.RWMutex{}
+var (
+	scopes          = make(map[string]*Scope)
+	defaultHandlers []ScopeHandlerCallbackFunc
+	lock            sync.RWMutex
+)
+
+// ScopeHandlerCallbackFunc is a callback type for the handler called from Fatal*, Error*, Warn*, Info* and Debug*
+// function calls.
+type ScopeHandlerCallbackFunc func(
+	level Level,
+	scope *Scope,
+	ie *structured.Error,
+	msg string,
+	fields []zapcore.Field)
+
+// RegisterDefaultHandler registers a scope handler that is called by default from all scopes. It is appended to the
+// current list of scope handlers.
+func RegisterDefaultHandler(callback ScopeHandlerCallbackFunc) {
```

Comment:
nit: private function?

Reply:
I care a decent amount about making everything in log that can be private, private. To avoid confusion

Reply:
done.